deployment/ec2/amis.py

```diff
@@ -0,0 +1,35 @@
+"""Helper functions to handle EC2 related operations with Boto"""
+
+import boto
+import logging
+
+
+LOGGER = logging.getLogger('mmw')
+LOGGER.addHandler(logging.StreamHandler())
+LOGGER.setLevel(logging.INFO)
+
+
+def _prune_ami(ami):
+    """Actually deregister AMI and its associated snapshot"""
+    LOGGER.info('Identified that [%s] is eligible to be pruned..', ami.id)
+
+    ami.deregister(delete_snapshot=True)
+
+    LOGGER.info('Deregistered [%s] and snapshot [%s]..', ami.id,
+                ami.block_device_mapping['/dev/sda1'].snapshot_id)
+
+
+def prune(mmw_config, service, keep, aws_profile):
+    """Filter owned AMIs by service, environment, and count"""
+    stack_type = mmw_config['StackType']
+
+    conn = boto.connect_ec2(profile_name=aws_profile)
+    images = conn.get_all_images(owners='self', filters={
+        'tag:Service': service,
+        'tag:Environment': stack_type
+    })
+
+    if len(images) > keep:
+        map(_prune_ami, reversed(sorted(images[0:len(images) - keep])))
```

Comment:
Had to put this to the test:

```
ami-cdce3ca6 - mmw-worker-1434571992-d7d3da6c46f0655b700b00530fdf3ec0585e11e4
ami-17e31e7c - mmw-worker-1434643973-797e66613623850c32509d7b8cf9597b04dc2307
ami-abdd20c0 - mmw-worker-1434645344-2b4056d1c3b94138622f2f080253868da11ccc1f
ami-21da274a - mmw-worker-1434648388-dbb3d148d9242a23a4c1fc8fb480e24bab4257fd
ami-efd42984 - mmw-worker-1434649432-4d2647e62c08fd4975a3d564eed6fd3269b55825
ami-43d22f28 - mmw-worker-1434651785-9b5384dcb80ac368b59d8ce2d7a8d1cf8ab7afa4

Old:

Identified that [ami-cdce3ca6] is eligible to be pruned..
Deregistered [ami-cdce3ca6] and snapshot [snap-33ccca72]..

New:

Identified that [ami-43d22f28] is eligible to be pruned..
Deregistered [ami-43d22f28] and snapshot [snap-21b19357]..
Identified that [ami-efd42984] is eligible to be pruned..
Deregistered [ami-efd42984] and snapshot [snap-dfdd53ac]..
Identified that [ami-21da274a] is eligible to be pruned..
Deregistered [ami-21da274a] and snapshot [snap-26f08147]..
Identified that [ami-abdd20c0] is eligible to be pruned..
Deregistered [ami-abdd20c0] and snapshot [snap-bbb720ca]..
Identified that [ami-17e31e7c] is eligible to be pruned..
Deregistered [ami-17e31e7c] and snapshot [snap-bd97ebd9]..
```

Here, "old" represents:

```python
map(_prune_ami,
    list(sorted(images, key=lambda i: i.creationDate))[0:len(images) - keep]) 
```

And "new" represents:

```python
map(_prune_ami,
    list(reversed(sorted(images, key=lambda i: i.creationDate)))[0:keep])
```

So, I think there were two bugs with the original:

1. There is no need for `reversed`
2. `sorted` needs to be explicitly told to use `name` or `createdDate` to sort properly

Should be addressed in https://github.com/WikiWatershed/model-my-watershed/commit/1bab4920405a42fb9b712caf96489f0985980a23.